libraries/chain/include/eosio/chain/fork_database.hpp

```diff
@@ -95,15 +98,77 @@ namespace eosio::chain {
 
       void mark_valid( const bsp& h );
 
-      static const uint32_t magic_number;
-
-      static const uint32_t min_supported_version;
-      static const uint32_t max_supported_version;
-
    private:
       unique_ptr<fork_database_impl<bsp>> my;
    };
 
-   using fork_database_legacy = fork_database<block_state_legacy_ptr>;
-   
+   using fork_database_legacy_t = fork_database_t<block_state_legacy_ptr>;
+   using fork_database_if_t     = fork_database_t<block_state_ptr>;
+
+   /**
+    * Provides thread safety on fork_database_t and provide mechanism for opening the correct type
+    * as well as switching from legacy to instant-finality.
+    */
+   class fork_database {
+      mutable std::recursive_mutex m;
```

Comment:
I'm not a fan of recursive mutexes. However, currently controller uses nested calls of `fork_database.apply`. We can look at refactoring those so that no `apply` calls functions which also calls `apply` but that will take some work.